core/state/state_object.go

```diff
@@ -197,7 +216,7 @@ func (s *stateObject) GetCommittedState(key common.Hash) common.Hash {
 	// If the snapshot is unavailable or reading from it fails, load from the database.
 	if s.db.snap == nil || err != nil {
 		start := time.Now()
-		tr, err := s.getTrie()
+		tr, err := s.getTrie(true)
```

Comment:
It's probably problematic?

- Prefetcher is enabled if the associated snapshot layer is available
- Retrievals from snapshot can succeed if the requested item is covered 
- Retrievals from snapshot can fail if the requested item is not covered yet(under the generation)
- The scenario is possible a few data retrieval through snapshot succeed and some prefetching tasks have been scheduled; then a retrieval fails and the prefetch'd trie is abandoned

The witness of the trie could be incomplete

Reply:
The parameter to `getTrie` determines whether or not it will attempt to load the trie from the prefetcher, so it will not stop the prefetcher prematurely here.

Afaict the changes that need to be made for witness-gathering-enabled path are:
* if snap is enabled, prefetch storage access lists on each call to `GetCommittedState` ([i.e.](https://github.com/ethereum/go-ethereum/pull/29719/files#diff-493fcc97ac5eefa37eef088731c482655afbae6c1ebaead72a88b7746897c606R215-R220))
* terminate the prefetcher synchronously at the start of the `IntermediateRoot` ([i.e.](https://github.com/ethereum/go-ethereum/pull/29719/files#diff-c3757dc9e9d868f63bc84a0cc67159c1d5c22cc5d8c9468757098f0492e0658cR896))
* later in `IntermediateRoot`: accumulate storage access lists from the trie of each mutated state object ([i.e.](https://github.com/ethereum/go-ethereum/pull/29719/files#diff-c3757dc9e9d868f63bc84a0cc67159c1d5c22cc5d8c9468757098f0492e0658cR1317))
* in statedb commit, iterate set of non-mutated state objects, if we can resolve a trie from the prefetcher, add its access list to the witness ([i.e.](https://github.com/ethereum/go-ethereum/pull/29719/files#diff-c3757dc9e9d868f63bc84a0cc67159c1d5c22cc5d8c9468757098f0492e0658cR1401)).

Reply:
> it will not stop the prefetcher prematurely here.

yes, it won't abort the prefetching, but the issue is the `trie` will be resolved from database and be linked to stateObject. The trie in prefetcher will never be retrieved later on.

Reply:
Ah I see.  Yeah, it's an issue.

Reply:
It does work as intended, but I agree that the intent was wrong :)

Reply:
I pushed a change to fix this. Whilst the optimization of loading *from the EVM excution into* the prefetcher is still lost (cleaner IMO); the new code forcefully overwrites the cached trie from the prefetcher on update if available. To make it less magical, I made this overwrite more explicit in the update code instead of relying on the cached trie magically getting swapped out based on some weird call sequence.